### PR TITLE
attribute operations

### PR DESCRIPTION
**Pull request by michaeldorman**

### CODE REVIEW

File: 03-attribute-operations.qmd [hunk 1]
```diff
@@ -371,56 +372,63 @@ Note that in this case, we need to set the index of `coffee_data` to the `name_l
...
+Often, we would like to create a new column based on already existing columns. For example, we want to calculate population density for each country. For this we need to divide a population column, here `pop`, by an area column, here `area_km2`. Note that we are working on a copy of `world` named `world2` so that we do not modify the original layer:
 
 ```{python}
 world2 = world.copy()
 world2['pop_dens'] = world2['pop'] / world2['area_km2']
 world2
 ```
 
-Unite / combine / concatenate string columns...
+To paste (i.e., concatenate) together existing columns, we can use the ordinary Python string operator `+`, as if we are working with individual strings rather than `Series`. For example, we want to combine the `continent` and `region_un` columns into a new column named `con_reg`, suing `':'` as a separator. Subsequesntly, we remove the original columns using `.drop`:
```
Comment (Nowosad) [inline 03-attribute-operations.qmd:383]: typo: suing

File: 03-attribute-operations.qmd [hunk 2]
```diff
@@ -371,56 +372,63 @@ Note that in this case, we need to set the index of `coffee_data` to the `name_l
...
 
-Rename...
+Renaming one or more columns can be done using the `.rename` method combined with the `columns` argument, which should be a dictionary of the form `old_name:new_name`. The following command, for example, renames the lengthy `name_long` column to simply `name`:
 
 ```{python}
 world2.rename(columns={'name_long': 'name'})
 ```
 
-Renaming all columns...
+To change all column names at once, by position, we assign a `list` of the "new" column names into the `.columns` property. The `list` must be of the same length as the number of columns (i.e., `world.shape[1]`). This is illustrated below, which outputs the same `world2` object, but with very short names:
```
Comment (Nowosad) [inline 03-attribute-operations.qmd:405]: I think `,by position,` can be removed.

File: 03-attribute-operations.qmd [hunk 3]
```diff
@@ -371,56 +372,63 @@ Note that in this case, we need to set the index of `coffee_data` to the `name_l
...
 names = sorted(world2.columns, reverse=True)
 world2 = world2[names]
 world2
 ```
 
-Dropping geometry and converting the `GeoDataFrame` into a `DataFrame`: 
+Each of these attribute data operations, even though they are defined in the `pandas` package and applicable to any `DataFrame`, preserve the geometry column and the `GeoDataFrame` class. Sometimes, however, it makes sense to remove the geometry, for example to speed-up aggregation or to export just the attribute data for statistical analysis. To go from `GeoDataFrame` to `DataFrame` we need to:
+
+1. Drop the geometry column
+1. Converting from `GeoDataFrame` into a `DataFrame`
```
Comment (Nowosad) [inline 03-attribute-operations.qmd:424]: Converting -> Convert

**Review by Nowosad — APPROVED** [review]

**Comment by michaeldorman** [comment]
Thanks @Nowosad !

**Comment by michaeldorman** [comment]
Corrected according to the comments

**Pull request merged**